windows-agent/internal/proservices/landscape/executor.go

```diff
@@ -254,61 +252,47 @@ func installFromURL(ctx context.Context, homeDir string, distro gowsl.Distro, ro
 	return nil
 }
 
-func download(w io.Writer, url string) (err error) {
-	defer func() {
-		if err != nil {
-			err = fmt.Errorf("could not download %q: %v", url, err)
-		}
-	}()
+func download(ctx context.Context, f io.Writer, url, checksum string) (err error) {
+	defer decorate.OnError(&err, "could not download %q", url)
 
 	//nolint:gosec // G107 expects url to be const, but we are reading it from Landscape.
 	resp, err := http.Get(url)
 	if err != nil {
 		return err
 	}
-	defer resp.Body.Close()
 
 	if resp.StatusCode >= 400 {
 		return fmt.Errorf("http request failed with code %d", resp.StatusCode)
 	}
 
-	if _, err := io.Copy(w, resp.Body); err != nil {
+	// Verify checksum and write file to disk
+	reader := io.TeeReader(resp.Body, f)
+	match, err := checksumMatches(ctx, reader, checksum)
+	if err != nil {
 		return err
 	}
+	if !match {
+		return errors.New("got unexpected checksum")
+	}
 
 	return nil
 }
 
-func checksumMatches(path, wantChecksum string) (err error) {
-	defer func() {
-		if err != nil {
-			err = fmt.Errorf("error checking checksum for: %q: %v", path, err)
-		}
-	}()
-
-	if wantChecksum == "" {
-		return nil
-	}
+func checksumMatches(ctx context.Context, reader io.Reader, wantChecksum string) (match bool, err error) {
+	defer decorate.OnError(&err, "error checking checksum for: %q", reader)
 
 	// Checksum of the rootfs
-	f, err := os.Open(path)
-	if err != nil {
-		return err
-	}
-	defer f.Close()
-
 	h := sha256.New()
-	if _, err := io.Copy(h, f); err != nil {
-		return err
+	if _, err := io.Copy(h, reader); err != nil {
+		return match, err
 	}
 	gotChecksum := fmt.Sprintf("%x", h.Sum(nil))
 
-	out, err := os.ReadFile(path)
-
 	// Compare checksums
-	if gotChecksum != wantChecksum {
-		return fmt.Errorf("checksums do not match (want: %s got: %s) (from %s)", wantChecksum, gotChecksum, string(out))
+	match = wantChecksum == "" || wantChecksum == gotChecksum
```

Comment:
Hum, please avoid ternary expression with assignement and check at the same time. This is prone to logical errors.

Also, with what you get below, you are not erroring out (only logging) if the checksums don’t match, but still logging it. This will potentially end up with a double logging of checksum mismatching. Also, I don’t think that printing the current downloaded checksum is meaningfull to the user.

I suggest that:
1. we don’t the checksum function at all if `wantChecksum == ""`. You will still need 
2. This function don’t log at all. Let the logging be done by the caller with returning an error in download() if match is false.
3. You need to `io.Copy` though on the caller to consume (as a no-op) the reader so that it’s written to a file, even if we don’t check the checksum (but at least, we don’t pretend and do a checksum for nothing).

Reply:
Its still useful to have the checksums logged during testing, so is a debug log here instead alright?

Reply:
I don’t see an explicit need TBH that the test doesn’t cover (getting the checksum in case of failure won’t really help debugging it if I’m correct), I think let’s remove it (as you did) and then, we can reintroduce as needed.